src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs

```diff
@@ -1581,6 +1581,199 @@ .maxstack  3
 }");
         }
 
+        [Fact]
+        public void RemoveRestoreNullableAtArrayElement()
+        {
+            var source0 =
+@"
+using System;
+class C
+{
+    public static void M()
+    {
+        var arr = new int?[] { 0 };
+        foreach (var s in arr)
+        {
+            Console.WriteLine(1);
+        }
+    }
+}";
+            // Remove nullable
+            var source1 =
+@"using System;
+class C
+{
+    public static void M()
+    {
+        var arr = new int[] { 0 };
+        foreach (var s in arr)
+        {
+            Console.WriteLine(1);
+        }
+    }
+}";
+            // Restore nullable
+            var source2 = source0;
+
+            var compilation0 = CreateCompilation(source0, options: TestOptions.DebugDll);
+
+            var v0 = CompileAndVerify(compilation0, emitOptions: EmitOptions.Default);
+            v0.VerifyIL("C.M", @"
+{
+  // Code size       56 (0x38)
+  .maxstack  4
+  .locals init (int?[] V_0, //arr
+                int?[] V_1,
+                int V_2,
+                int? V_3) //s
+  IL_0000:  nop
+  IL_0001:  ldc.i4.1
+  IL_0002:  newarr     ""int?""
+  IL_0007:  dup
+  IL_0008:  ldc.i4.0
+  IL_0009:  ldc.i4.0
+  IL_000a:  newobj     ""int?..ctor(int)""
+  IL_000f:  stelem     ""int?""
+  IL_0014:  stloc.0
+  IL_0015:  nop
+  IL_0016:  ldloc.0
+  IL_0017:  stloc.1
+  IL_0018:  ldc.i4.0
+  IL_0019:  stloc.2
+  IL_001a:  br.s       IL_0031
+  IL_001c:  ldloc.1
+  IL_001d:  ldloc.2
+  IL_001e:  ldelem     ""int?""
+  IL_0023:  stloc.3
+  IL_0024:  nop
+  IL_0025:  ldc.i4.1
+  IL_0026:  call       ""void System.Console.WriteLine(int)""
+  IL_002b:  nop
+  IL_002c:  nop
+  IL_002d:  ldloc.2
+  IL_002e:  ldc.i4.1
+  IL_002f:  add
+  IL_0030:  stloc.2
+  IL_0031:  ldloc.2
+  IL_0032:  ldloc.1
+  IL_0033:  ldlen
+  IL_0034:  conv.i4
+  IL_0035:  blt.s      IL_001c
+  IL_0037:  ret
+}");
+
+            var compilation1 = CreateCompilation(source1, options: TestOptions.DebugDll);
+            var compilation2 = compilation0.WithSource(source2);
+
+            var methodData0 = v0.TestData.GetMethodData("C.M");
+            var method0 = compilation0.GetMember<MethodSymbol>("C.M");
+            var generation0 = EmitBaseline.CreateInitialBaseline(ModuleMetadata.CreateFromImage(v0.EmittedAssemblyData), methodData0.EncDebugInfoProvider());
+
+            var method1 = compilation1.GetMember<MethodSymbol>("C.M");
+            var diff1 = compilation1.EmitDifference(
+                generation0,
+                ImmutableArray.Create(new SemanticEdit(SemanticEditKind.Update, method0, method1, GetEquivalentNodesMap(method1, method0), preserveLocalVariables: true)));
```

Comment:
>GetEquivalentNodesMap [](start = 98, length = 21)

I'd suggest changing this to `GetSyntaxMapFromMarkers` and use `MarkedSource` to construct the source. 
`GetEquivalentNodesMap` is old helper that won't match everything that needs to be matched to demonstrate that the compiler reuses all local slots as it should.

The first generation has the following locals:
```
                int?[] V_0, //arr
                int?[] V_1,
                int V_2,
                int? V_3) //s
```

What we want the test to demonstrate is that slots for all user defined locals and all long-lived temps are reused in the second generation, provided that they are syntactically mapped correctly.
Here you can see that didn't happen since `arr` and `s` have been allocated new slots.

Reply:
I replaced the helper. Why should we expect variable slots to keep if there are type changes?

---
In reply to: [308405395](https://github.com/dotnet/roslyn/pull/37096#discussion_r308405395) [](ancestors = 308405395)